falcon/api.py

```diff
@@ -206,7 +206,11 @@ def __call__(self, env, start_response):
                     # process_response when no error_handler is given
                     # and for whatever exception. If an HTTPError is raised
                     # remaining process_response will be executed later.
-                    self._call_resp_mw(middleware_stack, req, resp, resource)
+
+                    # NOTE(joarley): The condition will prevent that process_reponse
+                    # is called before self._compose_error_response on HTTPErrors
+                    if not isinstance(ex, HTTPError):
```

Comment:
Doing this causes a subtle breaking change in the case that `ex` is an instance of `HTTPError`. Before the change, if one of the response middleware methods (RMMs), itself, raised an error of type `HTTPStatus` or `HTTPError`, that will be caught and one more attempt will be made to execute the remaining RMMs. After this change, if the resource's responder raises an error, then one of the RMMs raises an error, it will simply go unhandled.

Admittedly, this highlights a weakness of the current design; IMO the framework should be able to handle any or all of the RMM raising instances of `HTTPStatus` or `HTTPError`, after which the exception is addressed and the remaining middleware methods executed. This would be more consistent and predictable.